src/inspect_ai/agent/_as_tool.py

```diff
@@ -41,9 +48,15 @@ def as_tool(agent: Agent, description: str | None = None, **agent_kwargs: Any) -
         )
 
     async def execute(input: str, *args: Any, **kwargs: Any) -> ToolResult:
-        # prepare state and call agent
+        # prepare state
         state = AgentState(messages=[ChatMessageUser(content=input, source="input")])
-        state = await agent(state, *args, **(agent_kwargs | kwargs))
+
+        # run the agent with limits
+        try:
+            with using_limits(limits):
+                state = await agent(state, *args, **(agent_kwargs | kwargs))
+        except LimitExceededError as ex:
```

Comment:
One thing to discuss is whether we actually need this try/catch anywhere but in a top level solver. The idea behind it is that we want to be able to score the intermediate message history but we won't have access to it in the case of an exception. For `as_tool()`, `run()` and `handoff()` though we aren't worried about the scoring (and in fact would specifically not want to propagate the message history as it would be partial)